### DISCUSSION TITLE
Private properties with Alpine

### DISCUSSION THREAD

**Discussion by Babelfisch** [alpinejs/alpine#3886] [1. Help] [answered]
Hello!

Is there a way to use Javascript _private properties_ within Alpine?

With this Javascript:

```javascript
class PrivateTest {
  #data = [];

  get items() {
    return this.#data;
  }
}
```

And this HTML:

```html
<div x-data="{test: new PrivateTest()}">
  <p>Item count: <span x-text="test.items.length"></span></p>
  <ol>
    <template x-for="item in test.items">
      <li x-text="item.name"></li>
    </template>
  </ol>
</div>
```

I get only this error:

```
Alpine Expression Error: Cannot access invalid private field (evaluating 'this.#data')
```

Any ideas how can I use private properties in my classes?

**Comment by SimoTod** [comment, marked as answer]
Not really. Alpine wraps all your objects in JS proxies to provide the reactivity and the `this` context is the proxy itself rather then the private class therefore the private variable is not accessible.
You could in theory use `Alpine.raw()` but then you would lose the reactivity (so it really depends if you are expecting `#data` to change or not).

Either
```js
class PrivateTest {
  #data = [];

  get items() {
    return Alpine.raw(this).#data; // Works but you are making your class dependant on Alpine + no reactivity :( 
  }
}
```
or
```html
<div x-data="{test: new PrivateTest()}">
  <p>Item count: <span x-text="Alpine.raw(test).items.length"></span></p>
  <ol>
    <template x-for="item in Alpine.raw(test).items">
      <li x-text="item.name"></li>
    </template>
  </ol>
</div>
<!-- Html more complex than it should be + no reactivity :( -->
```

**Reply by Babelfisch** [reply]
Thanks for your clarification. In that case I will not use private properties.

**Reply by DrSensor** [reply]
I wish there is a magic `$new(Class, ...args)` for this

**Reply by ekwoka** [reply]
I've been interested in looking at how Alpine could support Classes better (there are definitely some benefits to that path, though I feel like many people that use classes in JS do so incorrectly). There are some things that make the most clean solution (like being able to pass a function or class into x-data or Alpine.data) impossible (no way to actually KNOW that something will require `new` or should use `new`). May just be best to use Alpine.data and pass a function that instantiates the class. Or adding a `new` static method? or an alpine provided wrapper? or even a separate .data method for registering the component.

Either way, with how Alpine currently is expected to work (`this` refers to the alpine context with the component stack) is fundamentally incapable of supporting ACTUAL private fields.

I'm not sure if how it works could be redone to enable them while still offering everything that `this` is supposed to do, or what the minimum cost would be of supporting them.

I think for now, just don't use real private properties is probably the best. What private members offer over public is basically meaningless so it feels like supporting it wouldn't be worth any tradeoffs.

**Reply by DrSensor** [reply]
I'm interested on fixing this.
My understanding is that every object in x-data (e.g `x-data={obj:{}}`, not object produced by `x-data=createObject()`) is wrapped as ES Proxy.

Does Alpine reactivity works on deep object too? For example
```html
<div x-data='{ a: {} }'
  x-effect=console.log(a.b.c)>
 <button @click='(a.b ??= {}).c = 1'>
    init C
  </button>
</div>
```
If yes then introducing `$shallow` might be more appropriate.

**Reply by ekwoka** [reply]
What exactly do you mean by introducing `$shallow`?

What does that even mean?

And yes, the reactivity engine supports deep reactivity, including into Maps and Sets.

**Comment by ekwoka** [comment]
There is mostly just no reason to use private properties. Like what is the real concern? That you yourself access the properties you yourself made private?

But if you must, you can just use a closer for creating the component.

```js

Alpine.data('foo', () => {
  const privates = Alpine.reactive({ fizz: 'buzz' })
  return {
    get publicFizz() {
      return privates.fizz
    }
  }
})
```
